### PR TITLE
fix: stableswap lp minting for pool with tokens with different decimals.

### PR DESCRIPTION
**Pull request by desamtralized**
## Enhance StableSwap pools with improved decimal handling
This PR enhances the StableSwap implementation to better handle tokens with different decimal precisions. 

#### Main changes:
- Added core math functions for calculating D invariant with proper decimal normalization
- Implemented simulation functionality for stable swap pools with different decimal precision tokens
- Upgraded dependencies: "mantra-common-testing" (1.1.3 → 1.2.0), "mantra-dex-std", and added "mantrachain-std" and "neutron-std-derive"
- Enhanced the provide_liquidity function to include proper pool references
#### Lint fixes:
- Fixed numeric literals in tests to use consistent digit grouping with underscores
- Updated error handling in farm-manager and fee-collector tests with simpler error pattern matching

These changes improve the stability and precision of StableSwap operations, particularly for pools with mixed decimal precision tokens (e.g., 6, 12, and 18 decimals in the same pool). Tests have been added to verify correct LP token minting across different decimal configurations.

## Related Issues
https://github.com/MANTRA-Chain/mantra-dex/issues/145

---

## Checklist:

- [x] I have read [MANTRA's contribution guidelines](https://github.com/MANTRA-Chain/mantra-dex/blob/main/docs/CONTRIBUTING.md).
- [x] My pull request has a sound title and description (not something vague like `Update index.md`)
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation.
- [x] The code is formatted properly `just fmt`.
- [x] Clippy doesn't report any issues `just lint`.
- [x] I have regenerated the schemas if needed with `just schemas`.

## Summary by CodeRabbit

- **New Features**
  - Enhanced stable swap pool calculations with decimal normalization for token amounts, improving support for mixed decimal precisions.
  - Added new integration tests for stable swap liquidity provision, including edge cases and decimal handling.
  - Introduced a Python simulation for stable swap pool logic to validate LP token minting.

- **Bug Fixes**
  - Corrected numeric literal usage in test setup to prevent potential errors.

- **Refactor**
  - Simplified error pattern matching and attribute checks in test code for improved readability and maintainability.
  - Updated test setup to use computed values for token balances, enhancing clarity.

- **Chores**
  - Updated and removed various dependencies for improved project maintenance.
  - Added lint suppression attributes to address specific warnings without altering functionality.

- **Tests**
  - Expanded and clarified test coverage for stable swap pools, decimal normalization, and LP minting scenarios.

### CODE REVIEW

File: contracts/pool-manager/src/helpers.rs [hunk 1]
```diff
@@ -1612,6 +1653,79 @@ mod tests {
             }
         }
     }
+
+    #[test]
+    fn test_normalize_amount() {
+        // Test scaling up (6 decimals to 18 decimals)
+        let amount = Uint128::from(1_000_000u128); // 1.0 with 6 decimals
+        let normalized = normalize_amount(amount, 6, 18).unwrap();
+        assert_eq!(normalized, Uint128::from(1_000_000_000_000_000_000u128)); // 1.0 with 18 decimals
```
Comment (mantricjavier) [inline contracts/pool-manager/src/helpers.rs:1662]: When handling so many zeros it is better to have inconsistent decimal grouping and then add a clippy exception to the test, otherwise you get clippy complaining

File: contracts/pool-manager/src/tests/integration/swap.rs [hunk 2]
```diff
@@ -2669,18 +2669,18 @@ fn setup_3pool_different_decimals(
     // Default initial balances (300T for each token)
     let balances = initial_balances.unwrap_or_else(|| {
         vec![
-            Uint128::new(300_000_000_000_000_000000000000000000u128),
-            Uint128::new(300_000_000_000_000_000000000000000000u128),
-            Uint128::new(300_000_000_000_000_000000000000000000u128),
+            Uint128::new(300_000_000_000_000_000_000_000_000_000_000_u128),
```
Comment (mantricjavier) [inline contracts/pool-manager/src/tests/integration/swap.rs:2672]: Let's revert this how it was, it's easier for us humans to read 😄 . You just need to add `#[allow(clippy::inconsistent_digit_grouping)]` on top of the test

File: contracts/pool-manager/src/helpers.rs [hunk 3]
```diff
@@ -246,26 +250,85 @@
...
+            ))
+            .ok()
+    } else {
+        amount
+            .checked_mul(Uint128::from(
+                10u128.pow((to_decimals - from_decimals) as u32),
+            ))
+            .ok()
+    }
+}
```
[the remark this answers is not in the document]
Comment (mantricjavier) [reply contracts/pool-manager/src/helpers.rs:278]: The max decimals the contract would handle is 18 (say erc-20 assets), it should be fine @coderabbitai

File: contracts/pool-manager/src/helpers.rs [hunk 4]
```diff
@@ -246,26 +250,85 @@ fn calculate_d_core(amp_factor: &u64, deposits: &[Uint128], n_coins: Uint128) ->
...
+        let decimals = pool_info
+            .asset_denoms
+            .iter()
+            .position(|denom| denom == &coin.denom)
+            .map(|idx| pool_info.asset_decimals[idx])
+            .unwrap();
+
+        let normalized = normalize_amount(coin.amount, decimals, max_decimals)?;
+        normalized_deposits.push(normalized);
+    }
```
[the remark this answers is not in the document]
Comment (mantricjavier) [reply contracts/pool-manager/src/helpers.rs:327]: @coderabbitai There are checks before this function is triggered to check the deposits match the assets in the pool, it's fine.

File: contracts/pool-manager/src/helpers.rs [hunk 5]
```diff
@@ -975,56 +1039,127 @@
...
 ) -> Result<Option<Uint128>, ContractError> {
-    // Initial invariant
-    let d_0 = compute_d(amp_factor, old_pool_assets).ok_or(ContractError::StableInvariantError)?;
+    // Calculate the total deposit amount
+    let total_deposit_amount: Uint128 = old_pool_assets
+        .iter()
+        .zip(new_pool_assets.iter())
+        .map(|(old, new)| new.amount.checked_sub(old.amount).unwrap_or_default())
+        .sum();
+
```
[the remark this answers is not in the document]
Comment (mantricjavier) [reply contracts/pool-manager/src/helpers.rs:1116]: It's fine, vectors are sorted before calling this function.

**Review by mantricjavier — APPROVED** [review]
excellent 👏
